File: backend/utils/crypto.py
"""加密工具模块 - 使用 AES 对称加密，解密前端传递的加密密码"""

import base64
import json
from typing import Any

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend


# 密钥 - 必须与前端一致
SECRET_KEY = b"vue-python-fw-2024-secret-key-aes"


def _pad(data: bytes) -> bytes:
    """PKCS7 填充"""
    block_size = 16
    padding_length = block_size - (len(data) % block_size)
    padding = bytes([padding_length] * padding_length)
    return data + padding


def _unpad(data: bytes) -> bytes:
    """移除 PKCS7 填充"""
    padding_length = data[-1]
    return data[:-padding_length]


def encrypt(plain_text: str) -> str:
    """
    加密字符串（后端内部使用）

    Args:
        plain_text: 明文字符串

    Returns:
        加密后的 base64 字符串
    """
    iv = b"1234567890123456"  # 16 bytes IV
    cipher = Cipher(
        algorithms.AES(SECRET_KEY),
        modes.CBC(iv),
        backend=default_backend(),
    )
    encryptor = cipher.encryptor()
    encrypted = encryptor.update(_pad(plain_text.encode())) + encryptor.finalize()
    return base64.b64encode(encrypted).decode()


def decrypt(cipher_text: str) -> str:
    """
    解密字符串

    Args:
        cipher_text: 加密的 base64 字符串

    Returns:
        解密后的明文
    """
    iv = b"1234567890123456"
    encrypted_data = base64.b64decode(cipher_text)
    cipher = Cipher(
        algorithms.AES(SECRET_KEY),
        modes.CBC(iv),
        backend=default_backend(),
    )
    decryptor = cipher.decryptor()
    decrypted = decryptor.update(encrypted_data) + decryptor.finalize()
    return _unpad(decrypted).decode()


def decrypt_json(cipher_text: str) -> dict[str, Any]:
    """
    解密 JSON 字符串

    Args:
        cipher_text: 加密的 JSON 字符串

    Returns:
        解密后的字典
    """
    decrypted = decrypt(cipher_text)
    return json.loads(decrypted)

File: frontend/src/utils/crypto.ts
/**
 * 加密工具模块 - 使用 AES 对称加密
 * 前后端使用相同的密钥进行加密/解密
 */
import CryptoJS from 'crypto-js'

// 密钥 - 生产环境应使用环境变量
const SECRET_KEY = 'vue-python-fw-2024-secret-key-aes'

/**
 * 加密字符串
 * @param plainText - 明文字符串
 * @returns 加密后的 base64 字符串
 */
export function encrypt(plainText: string): string {
  const encrypted = CryptoJS.AES.encrypt(plainText, SECRET_KEY)
  return encrypted.toString()
}

/**
 * 解密字符串
 * @param cipherText - 加密的 base64 字符串
 * @returns 解密后的明文
 */
export function decrypt(cipherText: string): string {
  const decrypted = CryptoJS.AES.decrypt(cipherText, SECRET_KEY)
  return decrypted.toString(CryptoJS.enc.Utf8)
}

/**
 * 生成随机salt用于增强安全性
 */
export function generateSalt(): string {
  return CryptoJS.lib.WordArray.random(16).toString()
}

/**
 * 带salt的加密 - 更加安全
 * @param plainText - 明文
 * @param salt - 盐值
 */
export function encryptWithSalt(plainText: string, salt: string): string {
  const key = CryptoJS.PBKDF2(SECRET_KEY, salt, {
    keySize: 256 / 32,
    iterations: 1000,
  })
  const encrypted = CryptoJS.AES.encrypt(plainText, key)
  return encrypted.toString()
}

/**
 * 带salt的解密
 * @param cipherText - 密文
 * @param salt - 盐值
 */
export function decryptWithSalt(cipherText: string, salt: string): string {
  const key = CryptoJS.PBKDF2(SECRET_KEY, salt, {
    keySize: 256 / 32,
    iterations: 1000,
  })
  const decrypted = CryptoJS.AES.decrypt(cipherText, key)
  return decrypted.toString(CryptoJS.enc.Utf8)
}